File: app/lib/core/gestures.dart
/// Multi-touch gestures, as intents.
///
/// The phone recognises the *shape* of a gesture — three fingers went left, two
/// fingers pinched apart — and the computer decides what that means, because
/// the chord depends on the desktop: Alt+Tab switches windows on Linux and
/// Windows and does nothing on a Mac. So [ZrGesture] is what crosses the wire
/// (`{t:'gesture', g:'app-next'}`) and agent 1.9+ resolves it.
///
/// [fallbackChord] exists only for older agents: they don't know the verb, so
/// the phone sends the chord itself and gets the mapping right for the OS the
/// agent reported. It is a copy of the agent's table, deliberately small and
/// deliberately frozen — new gestures go in the agent, not here.
library;

class ZrGesture {
  /// three fingers left / right
  static const appNext = 'app-next';
  static const appPrev = 'app-prev';

  /// three fingers up / down
  static const overview = 'overview';
  static const showDesktop = 'show-desktop';

  /// two fingers flicked sideways — browser/file-manager history
  static const navBack = 'nav-back';
  static const navForward = 'nav-forward';

  /// pinch on the trackpad
  static const zoomIn = 'zoom-in';
  static const zoomOut = 'zoom-out';

  static const workspaceNext = 'workspace-next';
  static const workspacePrev = 'workspace-prev';
}

class ZrChord {
  const ZrChord(this.key, [this.mods = const []]);
  final String key;
  final List<String> mods;
}

const _linux = <String, ZrChord>{
  ZrGesture.appNext: ZrChord('tab', ['alt']),
  ZrGesture.appPrev: ZrChord('tab', ['alt', 'shift']),
  ZrGesture.overview: ZrChord('meta'),
  ZrGesture.showDesktop: ZrChord('d', ['meta']),
  ZrGesture.navBack: ZrChord('left', ['alt']),
  ZrGesture.navForward: ZrChord('right', ['alt']),
  ZrGesture.zoomIn: ZrChord('=', ['ctrl']),
  ZrGesture.zoomOut: ZrChord('-', ['ctrl']),
  ZrGesture.workspaceNext: ZrChord('right', ['ctrl', 'alt']),
  ZrGesture.workspacePrev: ZrChord('left', ['ctrl', 'alt']),
};

const _windows = <String, ZrChord>{
  ZrGesture.appNext: ZrChord('tab', ['alt']),
  ZrGesture.appPrev: ZrChord('tab', ['alt', 'shift']),
  ZrGesture.overview: ZrChord('tab', ['meta']), // Task View
  ZrGesture.showDesktop: ZrChord('d', ['meta']),
  ZrGesture.navBack: ZrChord('left', ['alt']),
  ZrGesture.navForward: ZrChord('right', ['alt']),
  ZrGesture.zoomIn: ZrChord('=', ['ctrl']),
  ZrGesture.zoomOut: ZrChord('-', ['ctrl']),
  ZrGesture.workspaceNext: ZrChord('right', ['ctrl', 'meta']),
  ZrGesture.workspacePrev: ZrChord('left', ['ctrl', 'meta']),
};

const _darwin = <String, ZrChord>{
  ZrGesture.appNext: ZrChord('tab', ['meta']),
  ZrGesture.appPrev: ZrChord('tab', ['meta', 'shift']),
  ZrGesture.overview: ZrChord('up', ['ctrl']),
  ZrGesture.showDesktop: ZrChord('f11'),
  ZrGesture.navBack: ZrChord('[', ['meta']),
  ZrGesture.navForward: ZrChord(']', ['meta']),
  ZrGesture.zoomIn: ZrChord('=', ['meta']),
  ZrGesture.zoomOut: ZrChord('-', ['meta']),
  ZrGesture.workspaceNext: ZrChord('right', ['ctrl']),
  ZrGesture.workspacePrev: ZrChord('left', ['ctrl']),
};

/// The chord to send to an agent too old to understand the intent. Null when
/// this OS has no sensible binding — better nothing than a wrong keystroke on
/// someone's desktop.
ZrChord? fallbackChord(String gesture, String hostOs) {
  final table = switch (hostOs) {
    'darwin' => _darwin,
    'windows' => _windows,
    _ => _linux,
  };
  return table[gesture];
}

File: public/app/js/gestures.js
// Multi-touch gestures, as intents.
//
// The phone recognises the SHAPE of a gesture — three fingers went left, two
// fingers pinched apart. The computer decides what it means, because the chord
// depends on the desktop: Alt+Tab switches windows on Linux and Windows and
// does nothing on a Mac. So `{t:'gesture', g:'app-next'}` is what crosses the
// wire and agent 1.9+ resolves it.
//
// The chord table below is only for older agents, which would drop a verb they
// don't know. It mirrors the agent's own table; new gestures go there, not here.
const ZRGestures = (() => {
  const G = {
    appNext: 'app-next', appPrev: 'app-prev',
    overview: 'overview', showDesktop: 'show-desktop',
    navBack: 'nav-back', navForward: 'nav-forward',
    zoomIn: 'zoom-in', zoomOut: 'zoom-out',
  };

  const linux = {
    'app-next': ['tab', ['alt']],
    'app-prev': ['tab', ['alt', 'shift']],
    'overview': ['meta', []],
    'show-desktop': ['d', ['meta']],
    'nav-back': ['left', ['alt']],
    'nav-forward': ['right', ['alt']],
    'zoom-in': ['=', ['ctrl']],
    'zoom-out': ['-', ['ctrl']],
  };
  const windows = Object.assign({}, linux, {
    'overview': ['tab', ['meta']], // Task View
  });
  const darwin = {
    'app-next': ['tab', ['meta']],
    'app-prev': ['tab', ['meta', 'shift']],
    'overview': ['up', ['ctrl']],       // Mission Control
    'show-desktop': ['f11', []],
    'nav-back': ['[', ['meta']],        // works in Safari, Chrome and Finder
    'nav-forward': [']', ['meta']],
    'zoom-in': ['=', ['meta']],
    'zoom-out': ['-', ['meta']],
  };

  // set from the welcome frame: which OS the agent runs, and whether it speaks
  // the gesture verb at all
  let host = { os: 'linux', resolves: false };
  function setHost(os, resolves) { host = { os: os || 'linux', resolves: !!resolves }; }

  // The frame to send for an intent — the verb when the agent understands it,
  // its chord otherwise, null when this OS has no sensible binding (better
  // nothing than a wrong keystroke on someone's desktop).
  function frame(id) {
    if (host.resolves) return { t: 'gesture', g: id };
    const table = host.os === 'darwin' ? darwin : host.os === 'windows' ? windows : linux;
    const chord = table[id];
    return chord ? { t: 'key', k: chord[0], mods: chord[1] } : null;
  }

  return { G, setHost, frame };
})();
